Multi party, networked offline heads

ch1bo opened the pull request:
This is a change I encountered when rebasing `raft-network` for #1720 and was useful back in the spike, but would have also been valuable in the `hydra-doom` use case.

Anyways, this is adding a `--offline-head-seed` argument to offline mode and fixes the "simulation" opened head to be deterministic across multiple instances, resulting that the nodes can and do talk to each other and consequently sign snapshots.

---

* [x] CHANGELOG updated
* [x] Documentation updated
* [x] Haddocks updated
* [x] No new TODOs introduced

ch1bo committed:
Offline mode: allow for multi-party heads

This simulates the corresponding initTx and commitTx transactions

ch1bo committed:
Update changelog and offline mode docs

ch1bo committed:
Drop offline sub-command on hydra-node

In favor of the --initial-utxo et al options

ch1bo committed:
Add a --offline-head-seed command line option

ch1bo committed:
Specify mult-party networked offline head as e2e test

ch1bo committed:
TMP: Trace all send/receive on L2 network

ch1bo committed:
hydra-cluster: Store keys in state dirctory of hydra-node

ch1bo committed:
Increase timeout for waiting on connected

Seemingly is the ouroboros-network stack quite lazy on reconnecting.

ch1bo committed:
Revert "TMP: Trace all send/receive on L2 network"

This reverts commit ba6f57086fb03b9c136cd302fea98ed2f73e35e7.

ch1bo committed:
Bump version of hydra-node

ch1bo committed:
Add IsString HeadSeed as a weeder root

noonio commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -429,20 +404,32 @@ instance Arbitrary ChainConfig where
+    (eitherReader $ left show . deserialiseFromRawBytesHex AsHeadSeed . BSC.pack)
+    ( long "offline-head-seed"
+        <> metavar "HEX"
+        <> help "Offline mode: Hexadecimal seed bytes to derive the offline head id from. Needs to be consistent across the hydra-node instances."
```
But is it hexadecimal? Elsewhere you wrote `test` as a valid value.

ch1bo replied:
The command line option must be hex because we parse it into the `ByteString`. I could have used `str` and use the `IsString` instance, but would need to add a rendering `HeadSeed -> String` function just for this case.

If you insist, I can switch to that?

noonio commented on hydra-tx/src/Hydra/Tx/HeadId.hs:
```diff
@@ -61,6 +61,9 @@ newtype HeadSeed = UnsafeHeadSeed ByteString
   deriving stock (Show, Eq, Ord, Generic)
   deriving (ToJSON, FromJSON) via (UsingRawBytesHex HeadSeed)
 
+instance IsString HeadSeed where
```
Wildly calls into question the value of "Safe ..." idiom at all 😅 😅 😅 😅 😅

ch1bo replied:
The idiom of `UnsafeXXX` is mostly about the unwrapping IIRC? Anyways, as long as any `String` can be represented as a `HeadSeed`, and we don't have anything restricting that right now in the general case .. in the Cardano-L1-specific case it must be a `TxId`, I think it's fine?

noonio replied:
yeah; i think it's fine too

i think `UnsafeXXX` is about smart constructors that only allow constructor for valid values. in this case, it _seemed_ you wanted only "hexadecimal strings"; but any string would do.

noonio reviewed (changes requested):
Seems great; aside from minor clarification around whether or not the head-seed actually must be hexadecimals :)

Seems like it's better if it's just an arbitrary string.

ch1bo committed:
Drop Sanchonet from supportedNetworks

The testnet is phased out and we should also consider dropping it from KnownNetworks

ffakenz commented on hydra-node/test/Hydra/OptionsSpec.hs:
```diff
@@ -300,13 +299,41 @@ spec = parallel $
+            { chainConfig =
+                Offline
+                  OfflineChainConfig
+                    { offlineHeadSeed = "\00\01"
```
why is so diff from the arg being passed?

ch1bo replied:
What do you mean? It's the same hex as on the command-line, just as a `Bytestring`

ffakenz commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -352,21 +332,16 @@ instance FromJSON ChainConfig where
 
 data OfflineChainConfig = OfflineChainConfig
-  { initialUTxOFile :: FilePath
+  { offlineHeadSeed :: HeadSeed
```
so, if i close an offline head then i need to prune persistence and restart the node with diff config?

for online heads this is not the case.
i would maintain consistency btw both ux.

wdyt?

ch1bo replied:
It would just load the head which is in the persistence. Like an online head.

ffakenz commented on hydra-cluster/test/Test/CardanoNodeSpec.hs:
```diff
@@ -18,7 +18,7 @@ import System.Directory (doesFileExist)
 import Test.Hydra.Cluster.Utils (forEachKnownNetwork)
 
 supportedNetworks :: [KnownNetwork]
-supportedNetworks = [Mainnet, Preproduction, Preview, Sanchonet]
```
why?

noonio replied:
it's been deleted it seems

![image](https://github.com/user-attachments/assets/e11c8ca5-9256-4cd9-84db-e9c56095cd40)

but i don't see any other information about it

ch1bo replied:
Yes, see the commit message.
